### ISSUE TITLE
Error with tensorflow_probability

### ISSUE THREAD

**Issue by APramov** [greta-dev/greta#265]
Hello, I wanted to start playing around with the greta package but I am running into the following issue when running the example code

`pkg$get_distribution("tensorflow_probability")$version`

namely, I get (showing it with the traceback below)
` Error in py_call_impl(callable, dots$args, dots$keywords) : 
  DistributionNotFound: The 'tensorflow_probability' distribution was not found and is required by the application` 

Traceback:

`1. pkg$get_distribution("tensorflow_probability")`
```
stop(structure(list(message = "DistributionNotFound: The 'tensorflow_probability' distribution was not found and is required by the application", 
    call = py_call_impl(callable, dots$args, dots$keywords), 
    cppstack = structure(list(file = "", line = -1L, stack = "C++ stack not available on this system"), class = "Rcpp_stack_trace")), class = c("Rcpp::exception", 
"C++Error", "error", "condition")))
```

I have seen from other posts that this is a problem that others face, however I am uncertain how to proceed. I installed greta from the development branch just now (version 0.3.0.9001) and my tf_version() is 1.10

I would be greateful for any help you could provide

UPDATE:

I  just ran 
`tensorflow::install_tensorflow(extra_packages='tensorflow_probability')`
as per suggestion from another (now closed) issue and I get the following error:
```

PackagesNotFoundError: The following packages are not available from current channels:

  - tensorflow_probability

Current channels:

  - https://conda.anaconda.org/conda-forge/win-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.continuum.io/pkgs/main/win-64
  - https://repo.continuum.io/pkgs/main/noarch
  - https://repo.continuum.io/pkgs/free/win-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/win-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/win-64
  - https://repo.continuum.io/pkgs/pro/noarch
  - https://repo.continuum.io/pkgs/msys2/win-64
  - https://repo.continuum.io/pkgs/msys2/noarch

```
Not sure what to do next. I am on Windows if that helps...

**Comment by goldingn** [comment]
Yeah, installing tensorflow into an environment that R is looking at is painful. Technically this is a pain point in the tensorflow and reticulate R packages (and it's *really hard* to find all the ways installing python packages fails, so it's not surprising their functionality isn't perfect), so we're limited in what we can do. Though it's affecting a lot of potential greta users, so I'm keen to try and find a foolproof set of instructions if we can.

Could you run the following commands for me, and I'll see what I can do to help?

```r
# install.packages("devtools") # (if you don't have it already)
devtools::install_github("greta-dev/greta")
greta:::check_tf_version("message")
reticulate::conda_list()
greta:::have_tf()
```

You could also try entering the following commands at a terminal (Tools > Terminal > New Terminal in a recent RStudio installation):

```
conda create r-tensorflow
conda activate r-tensorflow
conda install tensorflow
pip install tensorflow-probability
conda deactivate
```

**Comment by APramov** [comment]
Hi @goldingn , thanks a lot for helping me out with debugging this! I am really looking forward to trying greta on my data.

OK, so I followed you suggestions, bear with me, as the output is long.
First, I did these the two steps as you suggested. I 

- reinstalled devtools
- reinstalled greta
 

I continue what you suggested and here is the output from 

`greta:::check_tf_version("message")`

```
greta requires TensorFlow (>=1.10.0) and Tensorflow Probability (>=0.5.0), 
but TensorFlow Probability isn't installed. Use:
   install_tensorflow(method = "conda")
    reticulate::conda_install("r-tensorflow", "tensorflow-probability", pip = TRUE)
to install the latest version. 
```
At this stage I branch out from your suggestions and proceed to try and do what is suggested in the output below. In the first step by running ` install_tensorflow(method = "conda")`,  the installation works (a bunch of new packages is installed, but not probability one), but it warns me that I have an old version of conda (4.4.8) and the latest one is 4.6.4.  I try to update my conda by running `conda update -n base conda` in the anaconda terminal and it works. I check it and now my version is actually 4.6.3. Re-doing the ` install_tensorflow(method = "conda")` command doesn't give me any warning, so I guess things are ok.

Then I run     `reticulate::conda_install("r-tensorflow", "tensorflow-probability", pip = TRUE)` and I get the following error message:

```
C:\Github\XXX>for /F "skip=2 tokens=2,*" %A in ('reg query "HKEY_LOCAL_MACHINE\SOFTWARE\Wow6432Node\Microsoft\VisualStudio\14.0" /v InstallDir') do SET "VSINSTALLDIR=%B" 
ERROR: The system was unable to find the specified registry key or value.

C:\Github\XXX>if "" == "" (set "VSINSTALLDIR=" ) 

C:\Github\XXX>if "" == "" (
ECHO "Did not find VS in registry or in VS140COMNTOOLS env var - exiting"  
 exit 1 
) 
"Did not find VS in registry or in VS140COMNTOOLS env var - exiting"
Error: Error 1 occurred installing packages into conda environment r-tensorflow
```
I could not proceed further with this, so I come back to the next step of your suggestions. I run 
`reticulate::conda_list()` and this is what I get:
```          name                                                                                python
1  ANACON~1                                                C:\\ProgramData\\Anaconda3\\python.exe
2  r-tensorflow C:\\Users\\XXX\\Documents\\.conda\\envs\\r-tensorflow\\python.exe
```

(I put XXX to shorten the path)

Next step from you post is `greta:::have_tf()` which returns TRUE.

-----------------------------------------------------------------------------------------------------------

I now move on to the second block of your answer. I tried to run the commands from tools new terminal, but I get the following error (already on teh first command): `bash: conda: command not found`. My RStudio is version 1.1.463. 

So while the commands did not work in the terminal mode of Rstudio, I tried to run all of those commands in the anaconda prompt instead (with admin mode) - I don't know if that made any sense, but I thought I'd just try and that the output might help the debugging.

Funnily enough `pip install tensorflow-probability` gave the following output
```
Requirement already satisfied: tensorflow-probability in c:\programdata\anaconda3\lib\site-packages (0.5.0)
Requirement already satisfied: six>=1.10.0 in c:\programdata\anaconda3\lib\site-packages (from tensorflow-probability) (1.12.0)
Requirement already satisfied: numpy>=1.13.3 in c:\programdata\anaconda3\lib\site-packages (from tensorflow-probability) (1.16.1)
```
`conda install tensorflow` also worked within the anaconda prompt (but it just reinstalled the whole thing again).

Hope this helps! I really appreciate your help.

**Comment by goldingn** [comment]
OK, thanks for the detailed information, that's really helpful!

Could you please try the following at the anaconda prompt?

```
conda activate r-tensorflow
conda install tensorflow
pip install tensorflow-probability
conda deactivate
```

Then open R and try to load and use greta.

**Comment by APramov** [comment]
@goldingn  thanks! I am trying to try what you suggested, but after running the first line, the cmd automatically closes itself, I can't even see what the output message is. Do you have any idea how I can stop that?

**Comment by goldingn** [comment]
Huh, try this instead?

```
source activate r-tensorflow
conda install -n r-tensorflow tensorflow
pip install tensorflow-probability
source deactivate
```

**Comment by APramov** [comment]
Nick, just to let you know that it works now! I just ran the example on the greta intro webpage and it went all good. 

I will have to try to install it at my home PC too, but following your instructions, I managed to make it run. I had to update my anaconda installation and I when I ran ```
conda install -n r-tensorflow tensorflow
pip install tensorflow-probability
```
I had to download additional packages (and their dependencies) which were required by tensorflow-probability, which I did, and then it all worked.
Thanks a lot for the help!
